### PR TITLE
Re-add all SwiftSyntax rules and support

### PR DESCRIPTION
**Pull request by jpsim**
This reverts realm/SwiftLint#3107.

### CODE REVIEW

**Comment by jpsim** [comment]
@marcelofabri you can use this PR to iterate on the packaging issues discussed in realm/SwiftLint#3105.

**Comment by marcelofabri** [comment]
Thanks!

File: Source/SwiftLintFramework/Helpers/library_wrapper.swift [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+    var start = output.startIndex
+    var end = output.startIndex
+    var contentsEnd = output.startIndex
+    output.getLineStart(&start, end: &end, contentsEnd: &contentsEnd, for: start..<start)
+    let xcrunFindSwiftPath = String(output[start..<contentsEnd])
+    guard xcrunFindSwiftPath.hasSuffix("/usr/bin/swift") else {
+        return nil
+    }
+    let xcrunFindPath = xcrunFindSwiftPath.appending(pathComponent: "macosx")
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Helpers/library_wrapper.swift]: I had to change this from the original `library_wrapper` from SourceKitten ⚠️

Do I need to change something in the Linux code path?

Comment (norio-nomura) [reply Source/SwiftLintFramework/Helpers/library_wrapper.swift]: This line should be same with original like:
```suggestion
    let xcrunFindPath = xcrunFindSwiftPath.deleting(lastPathComponents: 3)
```

File: Source/SwiftLintFramework/Helpers/library_wrapper.swift [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    var xcodeDeveloperDir: String {
+        return appending(pathComponent: "Xcode.app/Contents/Developer")
+    }
+
+    var xcodeBetaDeveloperDir: String {
+        return appending(pathComponent: "Xcode-beta.app/Contents/Developer")
+    }
+
+    var usrLibDir: String {
+        return appending(pathComponent: "/usr/lib/swift")
```
Comment (norio-nomura) [inline Source/SwiftLintFramework/Helpers/library_wrapper.swift]: `macosx` should be appended at here:
```suggestion
        return appending(pathComponent: "/usr/lib/swift/macosx")
```

File: Source/SwiftLintFramework/Helpers/library_wrapper.swift [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+struct DynamicLinkLibrary {
+    fileprivate let handle: UnsafeMutableRawPointer
+
+    func load<T>(symbol: String) -> T {
+        if let sym = dlsym(handle, symbol) {
+            return unsafeBitCast(sym, to: T.self)
+        }
+        let errorString = String(validatingUTF8: dlerror())
+        fatalError("Finding symbol \(symbol) failed: \(errorString ?? "unknown error")")
```
Comment (norio-nomura) [inline Source/SwiftLintFramework/Helpers/library_wrapper.swift]: It produces a violation:
```suggestion
        queuedFatalError("Finding symbol \(symbol) failed: \(errorString ?? "unknown error")")
```

File: Source/SwiftLintFramework/Helpers/library_wrapper.swift [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+        // try all fullPaths that contains target file,
+        // then try loading with simple path that depends resolving to DYLD
+        for fullPath in fullPaths + [path] {
+            if let handle = dlopen(fullPath, RTLD_LAZY) {
+                return DynamicLinkLibrary(handle: handle)
+            }
+        }
+
+        fatalError("Loading \(path) failed")
```
Comment (norio-nomura) [inline Source/SwiftLintFramework/Helpers/library_wrapper.swift]:
```suggestion
        queuedFatalError("Loading \(path) failed")
```

File: Source/SwiftLintFramework/Helpers/library_wrapper.swift [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    guard let swiftPath = Exec.run("/usr/bin/which", "swift").string else {
+        return nil
+    }
+
+    if linuxSourceKitLibPath == nil && linuxFindSwiftenvActiveLibPath == nil &&
+       swiftPath.hasSuffix("/shims/swift") {
+        /// If we hit this path, the user is invoking Swift via swiftenv shims and has not set the
+        /// environment variable; this means we're going to end up trying to load from `/usr/lib`
+        /// which will fail - and instead, we can give a more useful error message.
+        fatalError("Swift is installed via swiftenv but swiftenv is not initialized.")
```
Comment (norio-nomura) [inline Source/SwiftLintFramework/Helpers/library_wrapper.swift]:
```suggestion
        queuedFatalError("Swift is installed via swiftenv but swiftenv is not initialized.")
```

**Comment by norio-nomura** [comment]
Can Carthage manage the SwiftSyntax submodule just like any other?

**Comment by jpsim** [comment]
I won't be trying to fix this any time soon.

**Pull request closed without merging**